src/deconstruction.ts

```diff
@@ -0,0 +1,722 @@
+import * as cdk from 'aws-cdk-lib';
+import { CfnResource, Tags } from 'aws-cdk-lib';
+import { CfnInclude } from 'aws-cdk-lib/cloudformation-include';
+import { IConstruct } from 'constructs';
+import * as reflect from 'jsii-reflect';
+import {
+  isConstruct,
+  isDataType,
+  isEnumLikeClass,
+  isSerializableInterface,
+  SchemaContext,
+  schemaForPolymorphic,
+} from './jsii2schema';
+
+export interface Tag {
+  Key: string;
+  Value: string;
+}
+
+export interface Override {
+  ResourcePath: string;
+  Update: { Path: string; Value: any };
+  Delete?: { Path: string };
+  RemoveResource?: boolean;
+}
+
+export interface DeclarativeResource {
+  Type: string;
+  Properties?: any;
+  Overrides?: Override[];
+  DependsOn?: string | string[];
+  Tags?: Tag[];
+}
+
+export function resolveType(fqn: string) {
+  const [mod, ...className] = fqn.split('.');
+  // eslint-disable-next-line @typescript-eslint/no-require-imports
+  const module = require(mod);
+
+  let curr = module;
+  while (true) {
+    const next = className.shift();
+    if (!next) {
+      break;
+    }
+    curr = curr[next];
+    if (!curr) {
+      throw new Error(`unable to resolve class ${className}`);
+    }
+  }
+
+  return curr;
+}
+
+export interface ParsedIntrinsic {
+  readonly name: string;
+  readonly value: any;
+}
+
+/**
+ * Parse an intrinsic into its name and its value.
+ *
+ * Returns `undefined` if the passed value does not look like an intrinsic.
+ *
+ * @example
+ * ```
+ * const parsed = tryParseIntrinsic({ "Fn::GetAtt": ["MyLambda", "Arn"] });
+ * parsed.name === "Fn::GetAtt";
+ * parsed.value === ["MyLambda", "Arn"]
+ * ```
+ */
+export function tryParseIntrinsic(input: any): ParsedIntrinsic | undefined {
+  if (typeof input !== 'object') {
+    return undefined;
+  }
+
+  if (Object.keys(input).length !== 1) {
+    return undefined;
+  }
+
+  const name = Object.keys(input)[0];
+  const value = input[name];
+  return { name, value };
+}
+
+export function tryResolveRef(value: any) {
+  const fn = tryParseIntrinsic(value);
+  if (!fn || fn.name !== 'Ref') {
+    return undefined;
+  }
+
+  return fn.value;
+}
+
+export function tryResolveGetAtt(value: any) {
+  const fn = tryParseIntrinsic(value);
+  if (!fn || fn.name !== 'Fn::GetAtt') {
+    return undefined;
+  }
+
+  return fn.value;
+}
+
+export interface DeconstructCommonOptions {
+  readonly stack: cdk.Stack;
+  readonly typeRef: reflect.TypeReference;
+  readonly key: string;
+  readonly value: any;
+}
+
+export interface DeconstructValueOptions extends DeconstructCommonOptions {
+  readonly optional: boolean;
+}
+
+export function deconstructValue(options: DeconstructValueOptions): any {
```

Comment:
This looks like it's mapping CFN values to JavaScript values.

You don't have to change it in this PR, it's something for later, but that's making me uncomfortable. I'd much rather have some modeling of the CFN type system that we control. Especially since at some point we're going to have to do a context-sensitive translation between types.

For example:

```ts
Resources:
  Res1:
    Type: AWS::S3::Bucket
    Properties:
      BucketName: Duration.seconds(5)   # { type: 'string', value: 'Duration.seconds(5)' }
  Res2:
    Type: aws-cdk-lib.aws_lambda.Function
    Properties:
      timeout: Duration.seconds(5)      # { type: 'decdk-static-call', module: 'aws-cdk-lib', class: 'Duration', method: 'seconds', arguments: [5] }
```

The occurrence of the same string ultimately need to parse differently based on context. In this case, I've invented some modeled type system that would look like this:

```ts
type DecdkValue = 
  | { type: 'string', value: string }
  | { type: 'object', values: Record<string, DecdkValue> }
  // ...
  | { type: 'decdk-static-call', module: string, class: string, method: string, arguments: DecdkValue[] }
  ;
```

In the current typing it (implicitly, due to lack of formalization) looks like this:

```ts
type DecdkValue = any;

// Everywhere we consume DecdkValues, we'll remember to treat a JavaScript 'string' as a CFN 'string' literal
// Everywhere we consume DecdkValues, we'll remember to treat a JavaScript 'object' as a CFN 'object' literal 
//    (this will be complicated by the fact that nearly every type in JavaScript is of type 'object' so we'll have to do a bunch of case analysis every time)

// How will we represent static method calls? 
//
// Well I don't know. I guess we'll either use:
interface StaticMethodCall { type: 'decdk-static-method-call'; module: string; ... }
// and remember to check every 'object' for a 'type' field, or we'll use:

class StaticMethodCall { ... }
// and remember to 'instanceof' every DecdkValue.
```

I'd much prefer the clear itemization of "a `DecdkValue` can be one of these 16 types", modeled by a union.

Reply:
After a little longer inspection, looks like I was wrong. In the case of the static method call, looks like `deconstructValue` would have immediately called the static method and return the actual value, instead of returning a symbolic representation of the value. It looks like `deconstructValue` is doing parsing + evaluating at the same time.

Ultimately, parsing to an intermediary representation, doing some potential transforms on that intermediary representation, and THEN doing evaluation would probably yield a more flexible and easier to maintain design.

Leaving my previous (misguided) comment in place because it does describe the situation I would want to get to.

Reply:
Yes, I get the idea. I still need to think a bit more on how to implement that in a clean an principled way. Leaving it like this for now.